### PR TITLE
fix: update formparser for new starlette

### PR DESCRIPTION
**Pull request by sauyon**
Should fix the formparser for starlette==0.25.

@ssheng I've only done basic testing with a simple test, and there's one minor issue below, this may need polishing or we may want to just quickly replace it.

### CODE REVIEW

File: src/bentoml/_internal/utils/formparser.py [hunk 1]
```diff
@@ -202,17 +310,21 @@ async def populate_multipart_requests(request: Request) -> t.Dict[str, Request]:
...
+        if "headers" not in scope:
+            scope["headers"] = []
+
         req = Request(scope)
         req._form = FormData([(field_name, data)])  # type: ignore (using internal starlette APIs)
-        if isinstance(data, bytes):
-            req._body = data
-        else:
+        if isinstance(data, str):
+            req._body = bytes(data, "utf-8")
```
Comment (sauyon) [inline src/bentoml/_internal/utils/formparser.py:326]: I've hard-coded this to be utf-8 for now; I think we might also want to set the encoding header, but I'm going to leave this as-is for now.

**Review by aarnphm — COMMENTED** [review]
there are some typing needs to get updated.

**Review by ssheng — DISMISSED** [review]

**Comment by ssheng** [comment]
I'm merging this because it is blocking other PRs. @aarnphm, please still give it a review.

File: src/bentoml/_internal/utils/formparser.py [hunk 2]
```diff
@@ -19,76 +20,230 @@
...
+    END = 5
+
+
+@dataclass
+class MultipartPart:
+    content_disposition: typing.Optional[bytes] = None
+    field_name: str = ""
+    data: bytes = b""
+    file: typing.Optional[UploadFile] = None
+    item_headers: typing.List[typing.Tuple[bytes, bytes]] = field(default_factory=list)
```
Comment (aarnphm) [inline src/bentoml/_internal/utils/formparser.py:43]:
```suggestion
    content_disposition: bytes | None = None
    field_name: str = ""
    data: bytes = b""
    file: UploadFile | None = None
    item_headers: list[tuple[bytes, bytes]] = field(default_factory=list)
```

File: src/bentoml/_internal/utils/formparser.py [hunk 3]
```diff
@@ -19,76 +20,230 @@
...
+class FormParser:
     def __init__(self, headers: Headers, stream: t.AsyncGenerator[bytes, None]) -> None:
         assert (
             multipart is not None
         ), "The `python-multipart` library must be installed to use form parsing."
-        self.headers: Headers = headers
+        self.headers = headers
         self.stream = stream
-        self.messages: t.List[t.Tuple[MultiPartMessage, bytes]] = list()
+        self.messages: typing.List[typing.Tuple[FormMessage, bytes]] = []
```
Comment (aarnphm) [inline src/bentoml/_internal/utils/formparser.py:65]:
```suggestion
        self.messages: list[tuple[FormMessage, bytes]] = []
```

File: src/bentoml/_internal/utils/formparser.py [hunk 4]
```diff
@@ -19,76 +20,230 @@
...
+        return FormData(items)
+
+
+class MultiPartParser:
+    max_file_size = 1024 * 1024
+
+    def __init__(
+        self,
+        headers: Headers,
+        stream: typing.AsyncGenerator[bytes, None],
```
Comment (aarnphm) [inline src/bentoml/_internal/utils/formparser.py:134]:
```suggestion
        stream: t.AsyncGenerator[bytes, None],
```

File: src/bentoml/_internal/utils/formparser.py [hunk 5]
```diff
@@ -19,76 +20,230 @@
...
+        *,
+        max_files: typing.Union[int, float] = 1000,
+        max_fields: typing.Union[int, float] = 1000,
```
Comment (aarnphm) [inline src/bentoml/_internal/utils/formparser.py:137]:
```suggestion
        max_files: int | float = 1000,
        max_fields: int | float = 1000,
```

File: src/bentoml/_internal/utils/formparser.py [hunk 6]
```diff
@@ -19,76 +20,230 @@
...
+    ) -> None:
+        assert (
+            multipart is not None
+        ), "The `python-multipart` library must be installed to use form parsing."
+        self.headers = headers
+        self.stream = stream
+        self.max_files = max_files
+        self.max_fields = max_fields
+        self.items: typing.List[typing.Tuple[str, typing.Union[str, UploadFile]]] = []
```
Comment (aarnphm) [inline src/bentoml/_internal/utils/formparser.py:146]:
```suggestion
        self.items: list[tuple[str, str | UploadFile]] = []
```

File: src/bentoml/_internal/utils/formparser.py [hunk 7]
```diff
@@ -19,76 +20,230 @@
...
+        self._current_files = 0
+        self._current_fields = 0
+        self._current_partial_header_name: bytes = b""
+        self._current_partial_header_value: bytes = b""
+        self._current_part = MultipartPart()
+        self._charset = ""
+        self._file_parts_to_write: typing.List[typing.Tuple[MultipartPart, bytes]] = []
+        self._file_parts_to_finish: typing.List[MultipartPart] = []
+        self._files_to_close_on_error: typing.List[SpooledTemporaryFile] = []
```
Comment (aarnphm) [inline src/bentoml/_internal/utils/formparser.py:155]:
```suggestion
        self._file_parts_to_write: list[tuple[MultipartPart, bytes]] = []
        self._file_parts_to_finish: list[MultipartPart] = []
        self._files_to_close_on_error: list[SpooledTemporaryFile] = []
```

File: src/bentoml/_internal/utils/formparser.py [hunk 8]
```diff
@@ -19,76 +20,230 @@
...
+                if message_type == FormMessage.FIELD_START:
+                    field_name = b""
+                    field_value = b""
+                elif message_type == FormMessage.FIELD_NAME:
+                    field_name += message_bytes
+                elif message_type == FormMessage.FIELD_DATA:
+                    field_value += message_bytes
+                elif message_type == FormMessage.FIELD_END:
+                    name = unquote_plus(field_name.decode("latin-1"))
+                    value = unquote_plus(field_value.decode("latin-1"))
```
Comment (aarnphm) [inline src/bentoml/_internal/utils/formparser.py:122]: undefined functions.

**Review by aarnphm — APPROVED** [review]

**Pull request merged**